### PR TITLE
types: optimize MyDecimal.ToFloat64

### LINKED ISSUE #25021: remove the intermediate String results when converting MyDecimal to Float
## Feature Request

**Is your feature request related to a problem? Please describe:**

When TiDB converts a MyDecimal to a Float, it first converts the MyDecimal to a String and then parses the String into a Float.
This method can generate lots of unnecessary intermediate results and affect Golang GC.

image
image

**Describe the feature you'd like:**

Implement a new method to let TiDB convert MyDecimal to Float directly to avoid generate unnecessary intermediate results.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by wzru**
### What problem does this PR solve?

Issue Number: close #25021  

Problem Summary:

When TiDB converts a MyDecimal to a Float, it first converts the MyDecimal to a String and then parses the String into a Float.
This method can generate lots of unnecessary intermediate results and affect Golang GC.

### What is changed and how it works?

What's Changed:

MyDecimal.ToFloat64() implementation

How it Works:

To calculate a float64 value directly using MyDecimal

### Check List 

Tests 

- Unit test

Side effects

- Performance regression
    - Consumes less CPU
    - Consumes less MEM

### Release note 

- No release note

### CODE REVIEW

**Review by lzmhhh123 — COMMENTED** [review]
Could you check the failed unit test and fix it?

**Review by wjhuang2016 — COMMENTED** [review]
We need a performance benchmark to accept this change.

File: types/mydecimal.go [hunk 1]
```diff
@@ -1084,12 +1085,35 @@ func (d *MyDecimal) FromFloat64(f float64) error {
 }
 
 // ToFloat64 converts decimal to float64 value.
-func (d *MyDecimal) ToFloat64() (float64, error) {
-	f, err := strconv.ParseFloat(d.String(), 64)
-	if err != nil {
-		err = ErrOverflow
+func (d *MyDecimal) ToFloat64() (f float64, err error) {
+	digitsInt := int(d.digitsInt)
```
Comment (guo-shaoge) [inline types/mydecimal.go:1089]: I am more worried about compatibility issues. Do we have a better way to cover this issue.

**Comment by wzru** [comment]
> We need a performance benchmark to accept this change.

Added a benchmark for MyDecimal.ToFloat64()
It seems like 10 times faster:
image

**Comment by guo-shaoge** [comment]
No need to merge master so many times. It will trigger CI rerun~

File: types/mydecimal.go [hunk 2]
```diff
@@ -1084,12 +1085,42 @@ func (d *MyDecimal) FromFloat64(f float64) error {
...
+		return
+	}
+	wordsInt := (digitsInt-1)/digitsPerWord + 1
+	wordIdx := 0
+	for i := 0; i < digitsInt; i += digitsPerWord {
+		x := d.wordBuf[wordIdx]
+		wordIdx++
+		f += float64(x) * pow10off81[(wordsInt-wordIdx)*digitsPerWord+81]
+		//f += float64(x) * math.Pow10((wordsInt-wordIdx)*digitsPerWord)
+	}
```
Comment (guo-shaoge) [inline types/mydecimal.go:1109]: remove comment.

File: types/mydecimal_test.go [hunk 3]
```diff
@@ -143,6 +143,8 @@ func (s *testMyDecimalSuite) TestToFloat(c *C) {
 		{"-123.45", -123.45},
 		{"0.00012345000098765", 0.00012345000098765},
 		{"1234500009876.5", 1234500009876.5},
+		{"1e39", 1e39},
+		{"1e-39", 1e-39},
 	}
```
Comment (guo-shaoge) [inline types/mydecimal_test.go:148]: Maybe more testcase?

File: types/mydecimal.go [hunk 4]
```diff
@@ -1084,12 +1085,42 @@ func (d *MyDecimal) FromFloat64(f float64) error {
...
+	fracStart := wordIdx
+	for i := 0; i < digitsFrac; i += digitsPerWord {
+		x := d.wordBuf[wordIdx]
+		wordIdx++
+		f += float64(x) * pow10off81[-digitsPerWord*(wordIdx-fracStart)+81]
+		//f += float64(x) * math.Pow10(-digitsPerWord*(wordIdx-fracStart))
```
Comment (guo-shaoge) [inline types/mydecimal.go:1115]: make 81 a constant is more clear.

**Comment by guo-shaoge** [comment]
reset LGTM

**Review by lzmhhh123 — APPROVED** [review]
LGTM

File: types/mydecimal.go [hunk 5]
```diff
@@ -1084,12 +1087,42 @@ func (d *MyDecimal) FromFloat64(f float64) error {
...
+		}
+		return
 	}
-	return f, err
+	wordsInt := (digitsInt-1)/digitsPerWord + 1
+	wordIdx := 0
+	for i := 0; i < digitsInt; i += digitsPerWord {
+		x := d.wordBuf[wordIdx]
+		wordIdx++
+		//equivalent to f += float64(x) * math.Pow10((wordsInt-wordIdx)*digitsPerWord)
```
Comment (guo-shaoge) [inline types/mydecimal.go:1109]:
```suggestion
		// Equivalent to f += float64(x) * math.Pow10((wordsInt-wordIdx)*digitsPerWord) .
```

**Review by guo-shaoge — APPROVED** [review]

**Pull request merged**